### PR TITLE
[HuggingFace] Performance tests HF Inference

### PR DESCRIPTION
**Pull request by bozheng2**
*Issue #, if available:*

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON Checklist
* When creating a PR:
- [ ] I've modified `src/config/build_config.py` in my PR branch by setting `ENABLE_EI_MODE = True` or `ENABLE_NEURON_MODE = True`
* When PR is reviewed and ready to be merged:
- [ ] I've reverted the code change on the config file mentioned above
#### Benchmark Checklist
* When creating a PR:
- [ ] I've modified `src/config/test_config.py` in my PR branch by setting `ENABLE_BENCHMARK_DEV_MODE = True`
* When PR is reviewed and ready to be merged:
- [ ] I've reverted the code change on the config file mentioned above

## Reviewer Checklist
* For reviewer, before merging, please cross-check:
- [ ] I've verified the code change on the config file mentioned above has already been reverted

*Description:*

*Tests run:*

*DLC image/dockerfile:*

*Additional context:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py [hunk 1]
```diff
@@ -1,5 +1,177 @@
...
+class TimeoutException(Exception):
     pass
+
+def timeout_handler(signum, frame):
+    raise TimeoutException
+
+@pytest.mark.model("bert-base-uncased-pt")
+@pytest.mark.processor("cpu")
+@pytest.mark.cpu_test
+def test_sm_trained_model_cpu():
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py:37]: Please use input `huggingface_pytorch_inference` to pass the URI for the HuggingFace PyTorch Inference image into this test, and add `cpu_only` to the inputs to only test CPU images.

The URI for the image that is built in a PR or a pipeline execution is automatically assigned to the variable `huggingface_pytorch_inference` whenever it is used as a parameter on any test function.

Please also change the name of this function to include `pytorch` in the name, since it is dedicated to PyTorch Inference images.

Also, please use pytest parametrization for tests that follow exactly the same workflow:
```suggestion
@pytest.mark.model("bert-base-uncased-pt")
@pytest.mark.processor("cpu")
@pytest.mark.cpu_test
@pytest.mark.parametrize("task,model", [
    ("fill-mask", BERT_BASE_UNCASED_PT_MODEL), ("fill-mask", ROBERTA_BASE_PT_MODEL)
])
def test_sm_trained_model_cpu():
```

Please also define the S3 URIs for images in variables common to all functions at the top of the script, so that they can be modified with a single code-change, in the future. Example:
```
BERT_BASE_UNCASED_PT_MODEL = "s3://aws-dlc-sample-models/huggingface_models/pytorch/bert-base-uncased-pytorch.tar.gz"
BERT_BASE_UNCASED_TF_MODEL = "s3://aws-dlc-sample-models/huggingface_models/tensorflow/bert-base-uncased-tensorflow.tar.gz"
ROBERTA_BASE_PT_MODEL = "s3://aws-dlc-sample-models/huggingface_models/pytorch/roberta-base-pytorch.tar.gz"
ROBERTA_BASE_TF_MODEL = "s3://aws-dlc-sample-models/huggingface_models/tensorflow/roberta-base-tensorflow.tar.gz"
```

This way we can have just 4 functions for different combinations of TF and PT + CPU and GPU.

Please also make sure that each test function has a different function name, as the test report would be generated incorrectly when two test functions share the same name in the same python test script.

File: test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py [hunk 2]
```diff
@@ -1,5 +1,177 @@
...
+def _test_sm_trained_model(instance_type, model, task, input_data, framework, device, accelerator_type=None):
+    ecr_image = ""
+    if framework == "pytorch" and device == "cpu":
+        ecr_image = "669063966089.dkr.ecr.us-west-2.amazonaws.com/beta-huggingface-pytorch-inference:1.7.1-transformers4.6.1-cpu-py36-ubuntu18.04-beta-tested-2021-06-18-16-02-51"
+    elif framework == "pytorch" and device == "gpu":
+        ecr_image = "669063966089.dkr.ecr.us-west-2.amazonaws.com/beta-huggingface-pytorch-inference:1.7.1-transformers4.6.1-gpu-py36-cu110-ubuntu18.04-beta-tested-2021-06-18-16-02-51"
+    elif framework == "tensorflow" and device == "cpu":
+        ecr_image = "669063966089.dkr.ecr.us-west-2.amazonaws.com/beta-huggingface-tensorflow-inference:2.4.1-transformers4.6.1-cpu-py37-ubuntu18.04-beta-tested-2021-06-18-16-27-00"
+    else :
+        ecr_image = "669063966089.dkr.ecr.us-west-2.amazonaws.com/beta-huggingface-tensorflow-inference:2.4.1-transformers4.6.1-gpu-py37-cu110-ubuntu18.04-beta-tested-2021-06-18-16-27-00"
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py:147]: Please remove these, as the ECR image URIs can be automatically picked up using pytest fixtures and passed into this function as an `ecr_image` input variable.

File: test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py [hunk 3]
```diff
@@ -1,5 +1,177 @@
...
+
+    hf_model = Model(
+        model_data=model,
+        role="SageMakerRole",
+        image_uri=ecr_image,
+        predictor_cls=Predictor,
+        env={'HF_TASK': task}
+    )
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py:155]: You will need to pass a `sagemaker_session` to the `Model` to allow it to actually create a Model and endpoint through the account whose `session` must be used.

Example: https://github.com/aws/deep-learning-containers/blob/master/test/sagemaker_tests/huggingface/inference/integration/sagemaker/test_sm_trained_model.py#L60

File: test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py [hunk 4]
```diff
@@ -1,5 +1,177 @@
...
+    predictor.deserializer = JSONDeserializer()
+    signal.signal(signal.SIGALRM, timeout_handler)
+    signal.alarm(1800)
+    latencies = []
+    LOGGER = logging.getLogger('timeout')
+    try:
+        for i in range(520):
+            start = time.time()
+            predictor.predict(data)
+            if i > 20:  # Warmup 20 iterations
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py:171]: Please make the number of warmup iterations a global variable in this script, so that it can be changed easily if needed. We may want to increase/decrease this in the future.

File: test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py [hunk 5]
```diff
@@ -1,5 +1,177 @@
...
+                latencies.append((time.time() - start) * 1000)
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py:172]: This will store the invocation latency from the requester to the endpoint and back.

This may be a useful metric to track, but it can vary drastically based on where the requester is based, and networks speeds on any particular day.

Can we instead gather endpoint metrics directly from SageMaker?

https://docs.aws.amazon.com/sagemaker/latest/dg/monitoring-cloudwatch.html#cloudwatch-metrics-endpoint-invocation

File: test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py [hunk 6]
```diff
@@ -1,5 +1,177 @@
...
+    except TimeoutException:
+        LOGGER.info("Timeout {}, {}, {}.".format(model, framework, device))
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py:174]: This needs to be either `LOGGER.warning`, or preferably `LOGGER.error`. Is there any reason to treat timeouts as a non-issue, here?

What is the expected behavior when there is a timeout error? This seems to proceed as usual and report the list of latencies back to the parent function. It would be better to have the test function fail right here with an `Exception`, so that the test job fails.

Otherwise, as long as the mean stays below 350 ms, we would mark it as successful even if there is an endpoint failure in the middle of the test.

File: test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py [hunk 7]
```diff
@@ -1,5 +1,177 @@
...
+@pytest.mark.cpu_test
+def test_sm_trained_model_cpu():
+    instance_type = "ml.m5.xlarge"
+    model = "s3://aws-dlc-sample-models/huggingface_models/pytorch/bert-base-uncased-pytorch.tar.gz"
+    task = "fill-mask"
+    input_data = {"inputs": "Hello I'm a [MASK] model."}
+    framework = "pytorch"
+    device = "cpu"
+    latencies = _test_sm_trained_model(instance_type, model, task, input_data, framework, device)
+    assert np.average(latencies) <= 350
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py:45]: Please configure the expected latency value into the following file:
https://github.com/aws/deep-learning-containers/blob/master/src/benchmark_metrics.py

and get the variable from the file in a similar way to https://github.com/aws/deep-learning-containers/blob/master/test/dlc_tests/benchmark/sagemaker/tensorflow/training/test_performance_tensorflow_sm_training.py#L10-L16.

We may have different latency values based on Transformers versions, or even TF or PT framework versions, and it is better to be prepared for this.

Are both `bert-base-uncased` as well as `roberta-base` expected to have the same latency?

File: test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py [hunk 8]
```diff
@@ -1,5 +1,177 @@
-import pytest
+# Copyright 2019-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py:1]: Please change this to 2019-2021.

File: test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py [hunk 9]
```diff
@@ -1,5 +1,177 @@
...
+import pytest
+import sagemaker
+from sagemaker.model import Model
+from sagemaker.predictor import Predictor
+from sagemaker.serializers import JSONSerializer
+from sagemaker.deserializers import JSONDeserializer
+import time
+import numpy as np
+import signal
+import logging
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py:26]: nit: Please sort imports in the following order:
```
from __future__ import absolute_import
# 1 line-space

# imports for all default packages for python such as os, logger, sys, in lexicographical order
import os
import signal
import time
# 1 line-space

# "from" imports for all default packages for python in lexicographical order
from signal import alarm  # or anything similar, if it exists
# 1 line-space

# imports for all installed packages such as numpy, pytest, sagemaker, in lexicographical order
import numpy as np
import pytest
import sagemaker
# 1 line-space

# "from" imports for all installed packages in lexicographical order
from sagemaker.deserializers import JSONDeserializer
from sagemaker.model import Model
from sagemaker.predictor import Predictor
from sagemaker.serializers import JSONSerializer
# 1 line-space

# imports for all local scripts, if needed
import test.test_utils.ecr as ecr_utils
# 1 line-space

# "from" imports for all local scripts
from test.test_utils import LOGGER
```

File: test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py [hunk 10]
```diff
@@ -1,5 +1,177 @@
...
+        initial_instance_count=1,
+        instance_type=instance_type
+    )
+    data = input_data
+    predictor.serializer = JSONSerializer()
+    predictor.deserializer = JSONDeserializer()
+    signal.signal(signal.SIGALRM, timeout_handler)
+    signal.alarm(1800)
+    latencies = []
+    LOGGER = logging.getLogger('timeout')
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/huggingface/inference/test_performance_huggingface_inference.py:166]: Why does this need a separate `LOGGER` from the one defined in https://github.com/aws/deep-learning-containers/blob/master/test/test_utils/__init__.py#L22-L24 ?
